src/lib/mgba/websocket-client.ts

```diff
@@ -1,452 +1,349 @@
-    if (length <= MEMORY_CONSTANTS.SMALL_READ_THRESHOLD) {
-      return this.readBytesBulk(address, length)
+  async eval (code: string): Promise<{ result?: string, error?: string }> {
+    if (!this.connected || !this.ws) {
+      throw new Error('Not connected to WebSocket server')
     }
 
-    try {
-      // Try using readRange API - fastest for larger reads
-      const luaCode = `(function()
-        local data = emu:readRange(${address}, ${length})
-        local bytes = {}
-        for i = 1, #data do
-          bytes[i] = string.byte(data, i)
-        end
-        return bytes
-      end)()`
-
-      const response = await this.eval(luaCode)
-      if (response.error) {
-        throw new Error(`Failed to read ${length} bytes at 0x${address.toString(16)}: ${response.error}`)
+    return new Promise((resolve, reject) => {
+      const timeout = setTimeout(() => {
+        reject(new Error('Eval request timed out'))
+      }, 5000)
+
+      const handler = (message: SimpleMessage): boolean => {
+        if (message.command === 'eval') {
+          clearTimeout(timeout)
+
+          if (message.status === 'error') {
+            resolve({ error: message.data.join('\n') })
+          } else {
+            const resultData = message.data.join('\n')
+            // Try to parse JSON result, otherwise return as string
+            try {
+              const parsed = JSON.parse(resultData)
+              resolve({ result: JSON.stringify(parsed) })
+            } catch {
+              resolve({ result: resultData })
+            }
+          }
+          return true
+        }
+        return false
       }
 
-      return new Uint8Array(response.result as number[])
-    } catch (error) {
-      // Fallback to bulk read if readRange fails
-      console.warn(`readRange failed, falling back to bulk read: ${String(error)}`)
-      return this.readBytesBulk(address, length)
-    }
+      this.pendingEvalHandlers.push(handler)
+      const message = ['eval', code].join('\n')
+      this.ws!.send(message)
+    })
   }
 
   /**
-   * Read multiple bytes using optimized bulk Lua operations (FAST)
-   * This is 100x+ faster than individual byte reads
+   * Read bytes from memory - uses cached data if available, otherwise eval
    */
-  async readBytesBulk (address: number, length: number): Promise<Uint8Array> {
-    const luaCode = `(function() 
-      local r = {} 
-      for i = 0, ${length - 1} do 
-        r[i+1] = emu:read8(${address} + i) 
-      end 
-      return r 
-    end)()`
-
-    const response = await this.eval(luaCode)
-    if (response.error) {
-      throw new Error(`Failed to bulk read ${length} bytes at 0x${address.toString(16)}: ${response.error}`)
+  async readBytes (address: number, size: number): Promise<Uint8Array> {
+    // Check if we have this data cached from watching
+    const cacheKey = `${address}-${size}`
+    const cachedData = this.watchedMemoryData.get(cacheKey)
```

Comment:
what if we want to only read a part of memory that might be watched? like a single value of a pokemon data struct. then this address-size key store is not good enough. we want to be able to read parts of the watched data without having to send a direct read out